### PR TITLE
Design & implement elastic markable memory chunks

### PR DESCRIPTION
**Pull request by OlegPliss**
The heap is represented by two contiguous resizable memory chunks, one of which stores the data, the other stores a marking bit for each 64-bit aligned word of the data. The mark can be set, cleared and tested. `test-and-set` is an additional fused operation for frequently used sequence of test and set.

Marked words can be iterated in a given address range of the heap. The iteration is going to be used for testing & debugging, in recovery from marking stack overflow and in remembered set scan of young collections.  

The heap is going to be aggregated to VMState. It can be initialized, finalized and resized in place.

### CODE REVIEW

File: core/core.stanza [hunk 1]
```diff
@@ -999,6 +1000,193 @@ lostanza defn collect-garbage (size:long) -> long :
...
+  val index = bit-index(heap, p)
+  val address = bit-address(heap, index)
+  [address] = [address] & (~ bit-mask(index))
+  return 0
+
+lostanza defn clear-mark (heap:ptr<Heap>, start:ptr<?>, limit:ptr<?>) -> int :
+  check-range(heap, start, limit)
+
+  val start-bit-index = bit-index(heap, start)
+  val end-bit-index = bit-index(heap, limit - 1)
```
Comment (CuppoJava) [inline core/core.stanza:1187]: This doesn't actually make any difference, but would the following be a more accurate representation of your intent? This is based upon my understanding of what `bit-index` does.

```
val end-bit-index = bit-index(heap, limit - BYTES-IN-LONG)
```

File: core/core.stanza [hunk 2]
```diff
@@ -999,6 +1000,193 @@ lostanza defn collect-garbage (size:long) -> long :
...
+    if (bits & 0xFFL) == 0 :
+      bits = bits >> 8
+      p = p + (8 * BYTES-IN-LONG)
+    if (bits & 0xFL) == 0 :
+      bits = bits >> 4
+      p = p + (4 * BYTES-IN-LONG)
+    if (bits & 0x3) == 0 :
+      bits = bits >> 2
+      p = p + (2 * BYTES-IN-LONG)
+    if (bits & 1) != 0 :
```
Comment (CuppoJava) [inline core/core.stanza:1154]: Does this condition always return true? Based upon the logic in line 1137 and the shifts in the previous lines.

File: core/core.stanza (hunk 1 above)
Comment (OlegPliss) [reply core/core.stanza:1187]: I expected this question. Yes, the result is the same but 1 is shorter than BYTES-IN-LONG.

**Review by CuppoJava — APPROVED** [review]
Looks good to me!

**Pull request merged**